cpp/FileTransferAgent.cpp

```diff
@@ -398,48 +403,66 @@ RemoteStorageRequestOutcome Snowflake::Client::FileTransferAgent::uploadSingleFi
   m_FileMetadataInitializer.initEncryptionMetadata(fileMetadata);
   CXX_LOG_TRACE("Encryption metadata init done");
 
-  std::basic_iostream<char> *srcFileStream;
-  ::std::fstream fs;
-
-  if (m_uploadStream)
+  RemoteStorageRequestOutcome outcome = RemoteStorageRequestOutcome::SUCCESS;
+  do
   {
-    srcFileStream = m_uploadStream;
-  }
-  else
- {
-    try {
-      fs = ::std::fstream(fileMetadata->srcFileToUpload.c_str(),
-                          ::std::ios_base::in |
-                          ::std::ios_base::binary);
-    }
-    catch(...)
+    if (retry > 0)
     {
-      std::string err= "Could not open source file " + fileMetadata->srcFileToUpload ;
-      throw SnowflakeTransferException(TransferError::FAILED_TO_TRANSFER, err.c_str());
+      unsigned long waitInMicroSecs = (retry << 1) * 100000 ; //First retry 200ms, Second retry 400ms, Third retry 800ms
```

Comment:
Also, this is the wrong way to do exponential backoff. Could you talk to @sfc-gh-mrui to see how he did exponential backoff for the spark connector ?

Reply:
Spark connector uses XP's backoff strategy (exponential backoff time with jitter).
1. start sleep time 3 second
2. max sleep time is 180 second,
3. Jitter factor is 0.5.
For example, the expected_sleep_time is 3, 6, 12, 24, etc
The sleep time is expected_sleep_time/2 + random(expected_sleep_time/2).

Reply:
Thanks @sfc-gh-mrui.
@sfc-gh-vreddy, what's missing here is the jitter.